### PR TITLE
DM-37033: update based on testing with the TMA in fall 2022

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/ts/mtmount/tma_telemetry_config_parser.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+                description = value
+            elif key_kind == "TCP_Publish":
+                if None in (name, units, description):
+                    raise RuntimeError(
+                        f"{name=!r}, {units=!r}, and {description=!r} cannot be empty; "
+                        f"key={key}; section {section.name}"
+                    )
+                try:
+                    # Ignore psp://192.168.209.10/PXIComm_NSV/CCW timestamp
+                    # which Tekniker insists have TCP_Publish = "TRUE"
```
Comment (teweitsai) [inline python/lsst/ts/mtmount/tma_telemetry_config_parser.py:327]: I am not sure it is allowed to expose the vendor's name (Tekniker) or not, which is supposed to be confidential.

File: python/lsst/ts/mtmount/mtmount_csc.py [hunk 2]
```diff
@@ -495,8 +516,11 @@ async def get_camera_cable_wrap_demand(self):
         )
 
         desired_tai = utils.current_tai() + self.config.camera_cable_wrap_advance_time
-        dt = rot_data.timestamp - desired_tai
+        dt = desired_tai - rot_data.timestamp
 
+        # Note: with recent rotator improvments, actual position and
```
Comment (teweitsai) [inline python/lsst/ts/mtmount/mtmount_csc.py:521]: improvements

File: python/lsst/ts/mtmount/mtmount_csc.py [hunk 3]
```diff
@@ -1448,6 +1536,23 @@ async def handle_power_state(self, reply):
             )
         else:
             await topic_info.topic.set_write(powerState=reply.powerState)
+        if (
+            reply.powerState == PowerState.FAULT
+            and self.summary_state == salobj.State.ENABLED
+        ):
+            # Send the CSC to fault if the CSC is enabled and:
+            # * the axis is azimuth and elevation
+            # * the axis is camera cable wrap and CCW following is enabled
```
Comment (teweitsai) [inline python/lsst/ts/mtmount/mtmount_csc.py:1545]: camera cable wrap or camera cable wrapper?

**Review by teweitsai — APPROVED** [review]
Looks good. I would suggest:

1. It might be better to have the "code reformat" in another PR to avoid the mixing of code format and new features.
2. For a big PR, it might be better to separate to multiple small PRs to let the review easier.
3. I should admit I could not follow all the code change and related logic in this big PR. Maybe you could go through all the code change again by yourself to make sure there is no mistake of logic.

Thanks!

File: doc/version_history.rst [hunk 4]
```diff
@@ -6,6 +6,43 @@
...
+
+    * Enable the oil supply system as part of enabling subsystems, now that the TMA can control it.
+    * Only issue the low-level heartbeat command when the CSC has control of the TMA.
+    * Give control of the TMA to the EUI when going to fault state, to avoid turning off the oil supply system and main power supply (which are slow to turn back on).
+    * Go to fault if telemetry stops arriving from the TMA.
+    * Go to fault if the azimuth, elevation, or camera cable wrap axis faults, with new error code ``CscErrorCode.AXIS_FAULT``.
+    * Log commands sent and command replies received at level 15 (halfway between info and debug).
+    * Add support for the GetActualSettings command. But don't use it yet, because it is broken in the TMA.
+    * Fix a bug in the ``monitor_telemetry_client`` method.
+    * Fix a bug in the camera cable wrap following code.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/version_history.rst:24]: The axis is called "camera cable wrap"

File: python/lsst/ts/mtmount/mock/controller.py [hunk 5]
```diff
@@ -460,22 +465,22 @@ async def put_axis_telemetry(self, system_id, tai):
...
-                # Torque is arbitrary; I have no idea
-                # what realistic values are.
-                torqueActual=actual.acceleration / 10,
+                actualPosition=actual.position,
+                demandPosition=target.position,
+                actualVelocity=actual.velocity,
+                demandVelocity=target.velocity,
+                # Torque is in Nm but I have little idea what realistic
+                # values are, so output something vaguely plausible
+                actualTorque=actual.acceleration / 10,
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/mtmount/mock/controller.py:483]: It is weird. It's just something vaguely plausible. I don't even know if the magnitude is correct.

File: python/lsst/ts/mtmount/mock/controller.py [hunk 6]
```diff
@@ -1179,6 +1184,18 @@ def do_both_axes_track_target(self, command):
...
+        """Handle the GET_ACTUAL_SETTINGS command.
+
+        Parameters
+        ----------
+        command : `Command`
+            The command.
+        """
+        task = asyncio.create_task(self.write_detailed_settings_applied())
+        timeout = 1
+        return timeout, task
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/mtmount/mock/controller.py:1197]: Good catch. Added to all do_ methods.

File: python/lsst/ts/mtmount/mock/oil_supply_system_device.py [hunk 7]
```diff
@@ -41,47 +42,97 @@ class OilSupplySystemDevice(BaseDevice):
...
+    * OilSupplySystemPowerCirculationPump:
+
+        * Only accepted in manual mode.
+        * Can only turn on if cooling is already on.
+        * Can only turn off if the main pump is off.
+
+    * OilSupplySystemPowerMainPump:
+
+        * Only accepted in manual mode.
+        * Can only turn on if cooling and the circulation pump are both on.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/mtmount/mock/oil_supply_system_device.py:65]: I'll reword. I don't want to imply that cooling is a pump.

File: python/lsst/ts/mtmount/tma_telemetry_config_parser.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/mtmount/tma_telemetry_config_parser.py:327]: This should be a matter of public record. I will risk it.

**Comment by r-owen** [comment]
The reformatting (isort) was done on a separate commit.

I agree it's a lot for one ticket. It's an accumulation of small improvements made while testing the TMA, rather than a focused change. I'm sorry it's not more compact.

**Pull request merged**